refactor, feat: provide per field typing & object serializer option

lihbr opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Description

Alright, so basically I wanted to experiment on top of your experiment (meta) to provide an object serializer option along the usual function serializer.

The idea being to allow the usage of the serialize function in a more... obvious way, as described here at #RichText: https://github.com/prismicio/prismic-reactjs/issues/92

Therefore with helpers or React it could look like this:

```typescript
import { asHTML } from "@priscmio/helpers";

asHTML(document.data.content, myLinkResolver, {
  heading1: ({ children }) => `<h2>${children.join("")}</h2>`
});
```

In order to achieve that I kinda reworked all the types to provide each field its own definitions, then fixed everything I broke.

I don't really know how to feel about it tbh, I'd like to see the object serializer becoming the norm over the function serializer (which is still handy for some particular edge cases, like creating a TOC or so)

All tests are still passing.

angeloashmore commented:
This looks great! I have some notes on the new concepts and performance changes.

---

**Reframe `buildFinalSerializer` as a serializer composer (e.g. `composeSerializers`)**

`buildFinalSerializer` effectively combines a default serializer with any number of serializers in order of priority.

If we think of `buildFinalSerializer` as just a way to combine serializers and fall back to other serializers if one does not return a non-nullish value, we can simplify the implementation. It becomes a serializer composer.

```ts
export const composeSerializers = <T>(
	...serializers: [
		RichTextFunctionSerializer<T>,
		...RichTextFunctionSerializer<T>[]
	]
) => {
	return (...args: Parameters<RichTextFunctionSerializer<T>>): T => {
		for (let i = 0; i < serializers.length; i++) {
			const res = serializers[i](...args);

			if (res != null) {
				return res;
			}
		}

		throw new Error(
			`Rich Text Node of type ${args[0]} does not have a serializer`,
		);
	};
};
```

Here, we throw if a node cannot be serialized. I think this is preferred over requiring a default serializer that must return a valid value since there could be cases where such a serializer doesn't exist.

For the following example, assume only headings nodes exist for simplicity. Each serializer listed below will return a valid value for the listed node types, and undefined for others.

```
Serializer A: `heading1`, `heading2`
Serializer B: `heading3`, `heading4`
Serializer C: `heading5`, `heading6`
```

By composing A, B, and C, we can cover all headings. There isn't a need to provide a fourth serializer that handles Heading 1-6. Instead, we throw.

This relies on run-time logic rather than static type analysis, so I understand if we want to avoid this approach and require a default serializer that always returns a value.

---

**Modularize object/map serializer from function serializer**

If we make the lower-level serializer functions less polymorphic, we can simplify their implementations. We can do this by writing everything as if only the function serializer exists. To support the object/map serializer, we can provide a function that converts a map to a function serializer.

This is exactly what you've done with `wrapObjectSerializer`, but it is now removed from `buildFinalSerializer` (now called `composeSerializers` as seen above). `composeSerializers` only accepts function serializers.

```ts
export const serializerFromMap = <T>(
	map: RichTextSerializerMap<T>,
): RichTextFunctionSerializer<T> => {
	return (...args: Parameters<RichTextFunctionSerializer<T>>): T => {
		const typeSerializer = map[ReversedNodeType[args[1].type]];

		if (typeSerializer) {
			return typeSerializer(...args);
		}
	};
};
```

(Note: This suffers from the same type problems as the existing `wrapObjectSerializer` due to `RichTextSerializerMap`'s [renamed from `RichTextObjectSerializer`] very specific function typing)

This allows higher level libraries to use `serializerFromMap` as needed. I think this extra verbosity is okay since most users will not be using this library directly.

---

**Performance decrease: `for` -> `Array.prototype.forEach`**

The change from `for` to `forEach` introduced a 1.5% reduction in number of ops/sec averaged over 6 runs. That's very little of course, but in this case the difference between `for` and `forEach` is minimal so I think we should go with the more performant solution.

https://github.com/prismicio/prismic-richtext/blob/f8824ca6dbf7267208d6249aa6eec5a376e789b4/src/asText.ts#L9-L13

---

**Performance increase: `"text" in richTextField[i]`**

The change from `if (richTextField[i].text)` to `if ("text" in richTextField[i])` in the above code snippet resulted in a 7% increase in number of ops/sec averaged over 6 runs.

The fastest implementation is:

```typescript
for (let i = 0; i < richTextField.length; i++) {
	if ("text" in richTextField[i]) {
		res += (i > 0 ? separator : "") + richTextField[i].text;
	}
}
```

This results in a TypeScript error since `richTextField[i]` is typed as `RTNode`, even if we did the `in` check before. If we restore `RTTextNode` (union of all RTNode types with a `text` field), we can coerce the type since we already know the `text` field exists.

```typescript
res += (i > 0 ? separator : "") + (richTextField[i] as RTTextNode).text;
```

To appease the type system, we could assign `richTextField[i]` to a const and assert the presence of the `text` field. This means we no longer need to coerce the type. This also means extra bytes in the package for no reason except to acquiesce to TypeScript. I'd vote for type coercion, but maybe there's a good reason we shouldn't?

---

**Interfaces vs type aliases**

For a library like this, I think we should prefer interfaces over type aliases (i.e. `interface Foo {}` vs `type Foo = {}`) as it allows developers to modify the types if needed. This could allow developers to fix incorrect types if we made an error or extend types if new API features are released. We would publish new versions in these instances of course, but it still allow users to be flexible.

We would only use `type Foo = {}` if we don't want a user to modify it, such as options to a function.

lihbr commented:
Alright!

- Refactored types over to `@prismicio/types`, published in v0.0.10: https://github.com/prismicio/prismic-types/commit/c9a2f598bb98aea458336aa4ff6391935c8642b9
- Refactored our two helpers functions `wrapMapSerializer` and `composeSerializers`: https://github.com/prismicio/prismic-richtext/pull/45/commits/b95f65b50027d649c2833110a26cead1b67d14af, https://github.com/prismicio/prismic-richtext/pull/45/commits/0c24e6b8eec94fe8c29e52f5c915addda579588a
- Discovered a nasty one while writing tests that lead `toTree` to mutate the provided rich rext field, fixed it there: https://github.com/prismicio/prismic-richtext/pull/45/commits/0674af8906a8566a23d60d2a11464979b2a50824, and added a test for it there: https://github.com/prismicio/prismic-richtext/pull/45/commits/8937a2463f526e3231e5fa4e5d6d15f2b325eb55
- Covered my changes, test coverage is at 100%: https://github.com/prismicio/prismic-richtext/pull/45/commits/6cbb9ab4d0c9be52c50f7b2f424b2cde695a5c7c, https://github.com/prismicio/prismic-richtext/pull/45/commits/7839146875f07b1369648b81fea2e446ed577005, https://github.com/prismicio/prismic-richtext/pull/45/commits/713681600b895239ec2d346a06715fa15ae74d7d
- Fixed the performance issues, didn't think this would add so much overhead 🙁 https://github.com/prismicio/prismic-richtext/pull/45/commits/2483ee5d02c94e163393ac6438e123f788121aeb

915 bytes gzipped :) (I'm cheating a bit because the NoteType enum is provided by `@prismicio/types` directly, but no more full reversed map)

angeloashmore commented:
This looks awesome! Thanks for catching that bug, too.

The ultimate size goal should be < 1kB on Bundlephobia, which I think will take into account `@prismicio/types` usage. I think we'll be under, or at least very close, with the current code.

I'm good to merge this in to the v2 branch if you don't have any other changes. 👍
